source/grid/grid_tools.cc

```diff
@@ -2088,6 +2088,159 @@ namespace GridTools
+              for ( unsigned int d=0; d<spacedim; ++d)
+                {
+                  minp[d] = std::min( minp[d], active_cells[i]->vertex(v)[d]);
+                  maxp[d] = std::max( maxp[d], active_cells[i]->vertex(v)[d]);
+                }
+
+        has_predicate = true;
+        BoundingBox < spacedim > bbox(std::make_pair(minp,maxp));
+        return bbox;
+      }
+    }
+  }
+
+
+
+  template < class MeshType >
+  std::vector< BoundingBox<MeshType::space_dimension> >
+  compute_mesh_predicate_bounding_box
+  (const MeshType &mesh,
+   const std::function<bool (const typename MeshType::active_cell_iterator &)> &predicate,
+   const unsigned int &refinement_level, const unsigned int &max_boxes)
+  {
+    // Algorithm brief description: beging with creating bounding boxes of all cells at
+    // refinement_level (and coarser levels if there are active cells) which have the predicate
+    // property. These are then merged
+
+    Assert( refinement_level <= mesh.n_levels(),
+            ExcMessage ( "Error: refinement level is higher then total levels in the triangulation!") );
+
+    const unsigned int spacedim = MeshType::space_dimension;
+    std::vector< BoundingBox < spacedim > > bounding_boxes;
+
+    // Creating a bounding box for all active cell on coarser level
+
+    for (unsigned int i=0; i < refinement_level; ++i)
+      for (typename MeshType::cell_iterator cell: mesh.active_cell_iterators_on_level(i))
+        {
+          bool has_predicate = false;
+          BoundingBox < spacedim > bbox =
+            internal::BoundingBoxPredicate::compute_cell_predicate_bounding_box <MeshType> (cell, predicate, has_predicate);
+          if (has_predicate)
+            bounding_boxes.push_back(bbox);
+        }
+
+    // Creating a Bounding Box for all cells on the chosen refinement_level
+    for (typename MeshType::cell_iterator cell: mesh.cell_iterators_on_level(refinement_level))
+      {
+        bool has_predicate = false;
+        BoundingBox < spacedim > bbox =
+          internal::BoundingBoxPredicate::compute_cell_predicate_bounding_box <MeshType> (cell, predicate, has_predicate);
+        if (has_predicate)
+          bounding_boxes.push_back(bbox);
+      }
+
+
+    // Part 1: merging neighbours
+    // This array stores the indices of arrays we have already merged
+    std::vector<unsigned int> merged_boxes_idx;
+    bool found_neighbors = true;
+
+    // We merge only nighbours which can be expressed by a single bounding box
+    while (found_neighbors)
+      {
+        found_neighbors = false;
+        for (unsigned int i=0; i<bounding_boxes.size()-1; ++i)
+          {
+            if ( std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),i) == merged_boxes_idx.end())
+              for (unsigned int j=i+1; j<bounding_boxes.size(); ++j)
+                if ( std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),j) == merged_boxes_idx.end()
```

Comment:
This is an expensive O(n^2) algorithm, right? How do you think this will work if we have 100'000 cells or more? This needs to be mentioned in the description of the algorithm.

Reply:
The idea is that one should choose a coarse refinement_level: for my use case I wasn't thinking of more than 10/20 bounding boxes but, as you point out, this algorithm has a high cost.

I will shortly push a different version where, by default, the merging part is avoided altogether.